mycpp/marksweep_gc_test.cc

```diff
@@ -0,0 +1,92 @@
+
+#include "mycpp/runtime.h"
+#include "vendor/greatest.h"
+
+TEST string_collection_test() {
+  Str *test_str = StrFromC("foo");
+  {
+    // NOTE(Jesse): This causes a crash when this gets compiled against the
+    // cheney collector w/ GC_EVERY_ALLOC.  I did verify it doesn't crash with
+    // the marksweep allocator but didn't want to figure out how to tell the
+    // build system to not compile these tests against the cheney collector
```

Comment:
OK I think we need to figure out something around `dbg` mode, and `gcevery` variant, which uses the Cheney collector

One thing we could do is change the default??   Like `marksweep` is the default?

Maybe we're ready to do that

Reply:
> Also not strictly related to this change, but another issue I've been thinking about is array vs. linked list for the root set
> 
> Most GCs I've seen seem to use a linked list for the root set, even though to me that seems slower
> 
> It avoids the bug you fixed where you exceed the maximum number of roots though! (or actually not fixed, since we still have a max number, it's just higher)
> 
> The problem though is that we modify the root set A LOT, every time `StackRoots` is called, which is twice in every function!
> 
> Another thing we could use benchmarks for: #1309

Yeah I think what we want is a block allocator with a freelist that allocates blocks of, for instance, 64 cache boundary aligned pointers with a `Next` member to another block.  When we pop enough pointers to clear an entire block it goes on a freelist.  When we push 63 pointers and need another block we take one from the freelist, or `malloc` another one.

ie:
```
struct StackRootBlock
{
  void ** Pointers[63];
  StackRootsBlock *Next;
};

struct StackRootsBlockFreelist
{
  StackRootsBlock *FirstFreeBlock;
};

class StackRoots
{
  ... push/pop functions .. whatever ..

  StackRootsBlock *CurrentBlock;
  StackRootsFreelist BlockFreelist;
};
```